### ISSUE TITLE
rustdoc: Inherent methods on trait objects are not indexed correctly

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#24137]
For example [searching for `downcast_ref`](http://doc.rust-lang.org/std/?search=downcast_ref) generates incorrect links.

**Comment by bluss** [comment]
I think this fixes it: (It passes the tests we have and fixes that bug).

But I wanted to ask because maybe there's a reason we don't use this already.

```diff
diff --git a/src/librustdoc/html/render.rs b/src/librustdoc/html/render.rs
index 65652a4..ad372f1 100644
--- a/src/librustdoc/html/render.rs
+++ b/src/librustdoc/html/render.rs
@@ -921,8 +921,8 @@ impl DocFolder for Cache {
                         let last = self.parent_stack.last().unwrap();
                         let did = *last;
                         let path = match self.paths.get(&did) {
-                            Some(&(_, ItemType::Trait)) =>
-                                Some(&self.stack[..self.stack.len() - 1]),
+                            Some(&(ref fqp, ItemType::Trait)) =>
+                                Some(&fqp[..fqp.len() - 1]),
                             // The current stack not necessarily has correlation
                             // for where the type was defined. On the other
                             // hand, `paths` always has the right
```

**Comment by ollie27** [comment]
This was fixed by #34752. https://doc.rust-lang.org/nightly/std/?search=downcast_ref

**Comment by alexcrichton** [comment]
Yay!
